UI/output-obj.cpp

```diff
@@ -0,0 +1,148 @@
+#include "output-obj.hpp"
+#include "obs-frontend-api.h"
+
+static void onEvent(enum obs_frontend_event event, void *param)
+{
+	OutputObj *obj = static_cast<OutputObj *>(param);
+
+	OBSSourceAutoRelease source;
+
+	switch (event) {
+	case OBS_FRONTEND_EVENT_STUDIO_MODE_ENABLED:
+	case OBS_FRONTEND_EVENT_PREVIEW_SCENE_CHANGED:
+		source = obs_frontend_get_current_preview_scene();
+		break;
+	case OBS_FRONTEND_EVENT_STUDIO_MODE_DISABLED:
+		source = obs_frontend_get_current_scene();
+		break;
+	case OBS_FRONTEND_EVENT_SCENE_CHANGED:
+		if (!obs_frontend_preview_program_mode_active())
+			source = obs_frontend_get_current_scene();
+
+		break;
+	default:
+		break;
+	}
+
+	obj->SetSource(source.Get());
+}
+
+OutputObj::OutputObj(const char *id, OBSData settings, enum OutputType type_)
+	: type(type_)
+{
+	output = obs_output_create(id, id, settings, nullptr);
+}
+
+OutputObj::~OutputObj()
+{
+	Stop();
+}
+
+OBSOutput OutputObj::GetOutput()
+{
+	return output.Get();
+}
+
+enum OutputType OutputObj::GetType()
+{
+	return type;
+}
+
+void OutputObj::SetType(enum OutputType type_)
```

Comment:
This code does not remove the frontend event callback if the type is switched from `Preview`. I recommend using this code:
```c++
{
	if (type == type_)
		return;

  bool active = Active();
  if (active)
    Stop();

	type = type_;

	if (active)
		Start();
}
```

This stops the output before changing type, causing the event callback to be removed as necessary.